// @frontegg/client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsClient = exports.EntitlementsClient = exports.IdentityClient = exports.HttpClient = exports.withAuthentication = exports.FronteggAuthenticator = exports.FronteggContext = exports.AuditsClient = void 0;
const authenticator_1 = require("./src/authenticator");
Object.defineProperty(exports, "FronteggAuthenticator", { enumerable: true, get: function () { return authenticator_1.FronteggAuthenticator; } });
const frontegg_context_1 = require("./src/components/frontegg-context");
Object.defineProperty(exports, "FronteggContext", { enumerable: true, get: function () { return frontegg_context_1.FronteggContext; } });
const middlewares_1 = require("./src/middlewares");
Object.defineProperty(exports, "withAuthentication", { enumerable: true, get: function () { return middlewares_1.withAuthentication; } });
const clients_1 = require("./src/clients");
Object.defineProperty(exports, "AuditsClient", { enumerable: true, get: function () { return clients_1.AuditsClient; } });
Object.defineProperty(exports, "EntitlementsClient", { enumerable: true, get: function () { return clients_1.EntitlementsClient; } });
Object.defineProperty(exports, "EventsClient", { enumerable: true, get: function () { return clients_1.EventsClient; } });
Object.defineProperty(exports, "HttpClient", { enumerable: true, get: function () { return clients_1.HttpClient; } });
Object.defineProperty(exports, "IdentityClient", { enumerable: true, get: function () { return clients_1.IdentityClient; } });
__exportStar(require("./src/clients/hosted-login"), exports);
