// @frontegg/client/dist/src/clients/identity/token-resolvers/access-token-services/services/tenant-access-token.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenantAccessTokenService = void 0;
const config_1 = require("../../../../../config");
const types_1 = require("../../../types");
const access_token_service_1 = require("./access-token.service");
class TenantAccessTokenService extends access_token_service_1.AccessTokenService {
    constructor(httpClient) {
        super(httpClient, types_1.tokenTypes.TenantAccessToken);
        this.httpClient = httpClient;
    }
    async getEntityFromIdentity(entity) {
        const { data } = await this.httpClient.get(`${config_1.config.urls.identityService}/resources/vendor-only/tenants/access-tokens/v1/${entity.sub}`);
        return {
            ...entity,
            roles: data.roles,
            permissions: data.permissions,
        };
    }
    async getActiveAccessTokenIdsFromIdentity() {
        const { data } = await this.httpClient.get(`${config_1.config.urls.identityService}/resources/vendor-only/tenants/access-tokens/v1/active`);
        return data;
    }
}
exports.TenantAccessTokenService = TenantAccessTokenService;
