// @frontegg/client/dist/src/clients/http/http-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = void 0;
const axios_1 = require("axios");
class HttpClient {
    constructor(authenticator, axiosConfig) {
        this.authenticator = authenticator;
        this.axios = axios_1.default.create(axiosConfig);
        this.setupRequestInterceptor();
    }
    setupRequestInterceptor() {
        this.axios.interceptors.request.use(async (request) => {
            var _a;
            await this.authenticator.validateAuthentication();
            request.headers.set('x-access-token', this.authenticator.accessToken);
            if (request.baseURL) {
                if ((_a = request.headers) === null || _a === void 0 ? void 0 : _a['frontegg-vendor-host']) {
                    request.baseURL = request.baseURL.replace('api', request.headers['frontegg-vendor-host']);
                }
            }
            return request;
        });
    }
    async get(url, headers) {
        return this.axios.get(url, { headers });
    }
    async post(url, data, headers) {
        return this.axios.post(url, data, { headers });
    }
    async put(url, data, headers) {
        return this.axios.put(url, data, { headers });
    }
    async delete(url, headers, data) {
        return this.axios.delete(url, { headers, data });
    }
    async patch(url, data, headers) {
        return this.axios.patch(url, data, { headers });
    }
}
exports.HttpClient = HttpClient;
